mlir/tools/rocmlir-gen/rocmlir-gen.cpp

```diff
@@ -1831,12 +1812,140 @@ static LogicalResult populateTensorFillLogic(mlir::OpBuilder &b,
   return success();
 }
 
+// Convert the async.launch/async.await pattern back to func.call.
+void undoAsyncLaunchPass(Operation *cloneFunc) {
+  SymbolTableCollection symbolTable;
+  auto walker = [&](Operation *op) {
+    OpBuilder builder(op);
+    if (auto launch = dyn_cast<async::LaunchOp>(op)) {
+      SymbolRefAttr calleeAttr = launch->getAttrOfType<SymbolRefAttr>("callee");
+
+      // Token is assumed to be the sole result of async.launch.
+      assert(launch->getNumResults() == 1);
+      auto aresults = launch->getResults();
+      mlir::Type tokenType = aresults.front().getType();
+
+      // Remove any token operands, at any position.
+      auto operands = launch->getOperands();
+      BitVector indicesToErase(launch.getNumOperands());
+      for (auto idx : llvm::seq<int>(0, launch.getNumOperands()))
+        if (!launch->getOperand(idx) ||
+            launch->getOperand(idx).getType() == tokenType)
+          indicesToErase.set(idx);
+      launch->eraseOperands(indicesToErase);
+      operands = launch->getOperands();
+
+      auto call = builder.create<func::CallOp>(op->getLoc(), calleeAttr,
+                                               TypeRange{}, operands);
+
+      call->moveBefore(op);
+      op->dropAllUses();
+      op->erase();
+      return WalkResult::interrupt();
+    }
+    if (auto launch = dyn_cast<async::AwaitOp>(op)) {
+      op->erase();
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  };
+  while (cloneFunc->walk(walker).wasInterrupted()) {
+  }
+}
+
+static void
+insertValidationCalls(const rock::Conv2dGenerator::Config &genConfig,
```

Comment:
Abstracted from the "if (hasValidation)" section in the original, with "clone" section added.